dispatch.go

```diff
@@ -78,13 +87,49 @@ type call struct {
 	ch             chan error
 	method         string
 	seqid          int
+	arg            interface{}
 	res            interface{}
 	errorUnwrapper ErrorUnwrapper
 	profiler       Profiler
 }
 
-func (c *call) Init() {
-	c.ch = make(chan error)
+func newCall(m string, arg interface{}, res interface{}, u ErrorUnwrapper, p Profiler) *call {
+	return &call{
+		ch:             make(chan error),
+		method:         m,
+		arg:            arg,
+		res:            res,
+		errorUnwrapper: u,
+		profiler:       p,
+	}
+}
+
+func (d *dispatch) callLoop() {
+	for {
+		select {
+		case <-d.stopCh:
+			for k, v := range d.calls {
+				v.ch <- EofError{}
+				delete(d.calls, k)
+			}
+			return
+		case c := <-d.callCh:
+			seqid := d.nextSeqid()
+			c.seqid = seqid
+			v := []interface{}{MethodCall, seqid, c.method, c.arg}
+			d.calls[c.seqid] = c
+			err := d.enc.Encode(v)
```

Comment:
Yep, that's the first thing I did: moving the encoder to channel-based locking on the transport so it's goroutine safe.